applications/FluidDynamicsBiomedicalApplication/custom_utilities/parabolic_profile_utilities.cpp

```diff
@@ -0,0 +1,241 @@
+    // Note that we check that the initial maximum levels are sufficient
+    double current_max_dist = char_length;
+    std::size_t current_max_levels = WallDistanceLevels / 2;
+    while (std::abs(current_max_dist - char_length) < 1.0e-12) {
+        // Calculate the parallel distance with the new number of levels
+        parallel_distance_settings["max_levels"].SetInt(2 * current_max_levels);
+        if (domain_size == 2) {
+            ParallelDistanceCalculationProcess<2>(rFluidModelPart, parallel_distance_settings).Execute();
+        } else if (domain_size == 3) {
+            ParallelDistanceCalculationProcess<3>(rFluidModelPart, parallel_distance_settings).Execute();
+        } else {
+            KRATOS_ERROR << "Unsupported domain size in " << rFluidModelPart.Name() << ". Found DOMAIN_SIZE " << domain_size << " in ProcessInfo container." << std::endl;
+        }
+
+        // Get the new maximum distance
+        current_max_dist = block_for_each<MaxReduction<double>>(rFluidModelPart.Nodes(), [](const NodeType& rNode){
+            return rNode.GetValue(WALL_DISTANCE);
+        });
+        rFluidModelPart.GetCommunicator().GetDataCommunicator().MaxAll(current_max_dist);
+    }
+
+    // Reset boundary values to zero
+    block_for_each(rWallModelPart.Nodes(), [](NodeType& rNode){
+        rNode.GetValue(WALL_DISTANCE) = 0.0;
+    });
+}
+
+void ParabolicProfileUtilities::ImposeParabolicInlet(
+    ModelPart& rModelPart,
+    const double MaxParabolaValue)
+{
+    // Impose the parabolic inlet profile
+    ImposeParabolicProfile(rModelPart, MaxParabolaValue);
+}
+
+void ParabolicProfileUtilities::ImposeParabolicInlet(
+    ModelPart& rModelPart,
+    const GenericFunctionUtility::Pointer MaxParabolaValue)
+{
+    // Impose the parabolic inlet profile
+    ImposeParabolicProfile(rModelPart, MaxParabolaValue);
+}
+
+template<class TInputType>
+void ParabolicProfileUtilities::ImposeParabolicProfile(
+    ModelPart &rModelPart,
+    const TInputType& rMaxParabolaValue)
+{
+    // Get time value from model part ProcessInfo
+    const auto& r_process_info = rModelPart.GetProcessInfo();
+    KRATOS_ERROR_IF_NOT(r_process_info.Has(TIME)) << "TIME is not present in '" << rModelPart.FullName() << "' ProcessInfo." << std::endl;
+    const double time = r_process_info[TIME];
+
+    // Get the maximum distance to the wall
+    const double max_dist = block_for_each<MaxReduction<double>>(rModelPart.Nodes(), [](NodeType& rNode){
+        return rNode.GetValue(WALL_DISTANCE);
+    });
+    KRATOS_ERROR_IF(std::abs(max_dist) < 1.0e-12) << "WALL_DISTANCE is close to zero." << std::endl;
+
+    // Impose the parabolic profile values
+    block_for_each(rModelPart.Nodes(), array_1d<double,3>(), [time, max_dist, &rMaxParabolaValue](NodeType& rNode, array_1d<double,3>& rTLSValue){
+        //Calculate distance for each node
+        const double wall_dist = rNode.GetValue(WALL_DISTANCE) < 0.0 ? 0.0 : rNode.GetValue(WALL_DISTANCE);
```

Comment:
Do negative `WALL_DISTANCE` make sense in this app? If not, i would check and set this in the `CalculateWallParallelDistance`.

Reply:
Indeed the boundary values are reset to zero when calculating the distance. In any case I'd keep it just in case as the performance impact is negligible.